### ISSUE TITLE
Schema has not been registered

### ISSUE THREAD

**Issue by gurdeepsharma-boffincoders** [typegoose/typegoose#633]
## Versions

- System: linux / macos / windows 
- NodeJS: 14.17
- Typescript: 4.5.2

- Compiler / Transpiler: tsc / ts-node / ts-jest 
- Typegoose(NPM): 0.0.0 
- Typegoose(GIT): commithash
- mongoose: 6.0.11 
- mongodb: 0.0.0
- nestjs / nestjs-typegoose: 0.0.0 / 0.0.0 
- grahpql / type-graphql: 0.0.0 / 0.0.0 

## What is the Problem?

Schema has not been registered

## Code Example

```ts
1. BaseUsers

        import { getModelForClass, prop } from "@typegoose/typegoose";
        export class BaseUsers {
          @prop()
          name: string;
          @prop({ unique: true })
          email: string;
          @prop()
          password: string;
          @prop()
          country: string;
          @prop()
          pin_code: string;
          @prop()
          mobile_number: string;
          @prop()
          can_be_deleted: boolean;
          @prop()
          avatar: string;
        }
        
        const BASE_USERS_MODEL = getModelForClass(BaseUsers, {
          schemaOptions: {
            collection: "users",
            timestamps: true,
          },
        });
        export default BASE_USERS_MODEL;

2. Clients

        import { getModelForClass, Index, prop } from "@typegoose/typegoose";
        export class SMTPSettings {
          @prop()
          host: string;
          @prop()
          email: string;
          @prop()
          from: string;
          @prop()
          password: string;
          @prop()
          port: string;
          @prop()
          isSSL: boolean;
        }
        export class SendGridSettings {
          @prop()
          host: string;
          @prop()
          email: string;
          @prop()
          from: string;
          @prop()
          password: string;
          @prop()
          port: string;
          @prop()
          isSSL: boolean;
        }
        
        @Index(
          { name: 1, alias_name: 1 },
          {
            unique: true,
          }
        )
        export class BaseClients {
          @prop()
          name: string;
        
          @prop()
          alias_name: string;
        
          @prop()
          logo: string;
        
          @prop()
          akamai_cp_code: string;
        
          @prop()
          is_smtp: boolean;
        
          @prop({ _id: false })
          stmp_settings?: SMTPSettings;
        
          @prop({ _id: false })
          sendgrid_settings?: SendGridSettings;
        }
        
        const BASE_CLIENTS_MODEL = getModelForClass(BaseClients, {
          schemaOptions: {
            collection: "clients",
            timestamps: true,
          },
        });
        export default BASE_CLIENTS_MODEL;

3. Roles

    import { getModelForClass, prop } from "@typegoose/typegoose";

    export class Roles {
      @prop()
      name: string;
    }

     const RolesModel = getModelForClass(Roles, {
      schemaOptions: {
        collection: "roles",
      },
    });
export default RolesModel;

4. Clients Association

import { getModelForClass, mongoose, prop, Ref } from "@typegoose/typegoose";
    import { BaseClients } from "./clients";
    import { Roles } from "./roles";
    import { BaseUsers } from "./users";
    
    export class ClientIdAssociation {
    
      @prop({ ref: () => BaseClients })
      client_id?: Ref<BaseClients>;

      @prop({ ref: () => BaseUsers })
      user_id: Ref<BaseUsers>;

      @prop({ ref: () => Roles, required: true })
      roles!: Ref<Roles>[];
    
      @prop({ type: mongoose.Types.ObjectId })
      allowed_cms: mongoose.Types.ObjectId[];
    
      @prop({ type: mongoose.Types.ObjectId })
      allowed_crm: mongoose.Types.ObjectId[];
    
      @prop({ type: mongoose.Types.ObjectId })
      allowed_uit: mongoose.Types.ObjectId[];
    
      @prop({ type: mongoose.Types.ObjectId })
      allowed_iq: mongoose.Types.ObjectId[];
    }
    
    const CLIENT_ID_ASSOCIATION_MODEL = getModelForClass(ClientIdAssociation, {
      schemaOptions: {
        timestamps: true,
        collection: "user_association_with_clients",
      },
    });
    export default CLIENT_ID_ASSOCIATION_MODEL;

My DB Config looks like this

    import {
      AnyParamConstructor,
      ReturnModelType
    } from "@typegoose/typegoose/lib/types";
    import Mongoose from "mongoose";
    
    class DatabaseConfig {
      public database: Mongoose.Connection;
      public connectDatabase = (): any => {
        const uri =
          "mongodb://localhost:27017" ??
          "";
        if (this.database) {
          return;
        }
        Mongoose.connect(uri);
        this.database = Mongoose.connection;
        this.database.once("open", async () => {
          console.log("Connected to database");
        });
        this.database.on("error", () => {
          console.log("Error connecting to database");
        });
      };
      disconnectDatabase = () => {
        if (!this.database) {
          return;
        }
        Mongoose.disconnect();
      };
    
      getModelForDb<T extends AnyParamConstructor<any>>(
        databaseName: string,
        model: ReturnModelType<T>
      ): ReturnModelType<T> & T {
        const db = Mongoose.connection.useDb(databaseName);
    
        const DbModel = db.model(
          model.modelName,
          model.schema
        ) as ReturnModelType<T> & T;
    
        return DbModel;
      }
    }
    
    export default new DatabaseConfig();


```

Now when I try to get the reference populated it gives me error for roles while the client_id and user_id is correctly populated. I am using the below code for getting the results.

 
```
    let client_id_association_model = databaseConfig.getModelForDb<
            typeof ClientIdAssociation
          >("ims_base_db", CLIENT_ID_ASSOCIATION_MODEL);
    
    
     let found_users = await client_id_association_model
            .find({
              user_id: new mongoose.Types.ObjectId("6103c782f67f5162439f7d8e"),
            })
            .populate(["roles","client_id"]);
```
Kindly advise where is the issue.

Packages Version
```
       "@typegoose/typegoose": "^9.2.0",
        "express": "^4.17.1",
        "mongoose": "^6.0.11"
```
![image](https://user-images.githubusercontent.com/74892090/142842843-6189817f-73c4-4a80-87e8-f4aa83b1cb0b.png)

---

**Comment by manojsethi** [comment]
Facing the same issue while using the generic model method to support multiple dynamic databases.

**Comment by hasezoey** [comment]
> Packages Version

next time please put the versions where listed in the template, not somewhere else

---

on what line does the error come from? also could this example code be more minified or is this the least it can be before triggering the issue?

**Comment by manojsethi** [comment]
Bascially this happens when we are using the dynamic connection

```
let client_id_association_model = databaseConfig.getModelForDb<
            typeof ClientIdAssociation
          >("ims_base_db", CLIENT_ID_ASSOCIATION_MODEL);
```

if we don't use the dynamic connection string and include the db name inside the connection string then it works fine.

So the issue comes when we use the dynamic connection string for the purpose of multitenancy, I am using your code written at this comment for the purpose of multitenancy

https://github.com/typegoose/typegoose/issues/158#issuecomment-564712074

**Comment by hasezoey** [comment]
this seems to be a mongoose problem, because typegoose didnt change anything related to that (from what i can recall)

> if we don't use the dynamic connection string and include the db name inside the connection string then it works fine.

what exactly do you mean with that? do you mean the actual connection string (like `mongodb://localhost:27017/theDB`) or something else?

---

also my comment in that referenced issue (https://github.com/typegoose/typegoose/issues/158#issuecomment-564712074) was for typegoose 6.x (which used mongoose 5.9 in the last 6.x version), which is relatively old, i also never used that approach myself or tested the function at runtime (probably, i dont recall if i did it or not), because that issue was about types

---

also looking closer at the error and code again, it looks like the error comes when calling the method `getModelForDb` with model `CLIENT_ID_ASSOCIATION_MODEL` which is class `ClientIdAssociation`, so i can infer that the error is about property `roles` from the class, which probably requires model `RolesModel` (class `Roles`) to be in the connection before the `ClientIdAssociation` class / model

TL:DR; try to add the model of the classes of `BaseClients`, `BaseUsers` and `Roles` before `ClientIdAssociation` to the required connection

**Comment by manojsethi** [comment]
Yes with the actual connection string I mean to say that we append the db name in connection string.

Well we have been using the approach in comment https://github.com/typegoose/typegoose/issues/158#issuecomment-564712074 with typegoose version 7.6.3 and mongoose version ~5.10.x. The combination worked beautifully but since the types/mongodb stub is now empty we tried to upgrade the packages but faced this roadblock. 

I am not sure about adding the models before ClientIDAssociation before because the flow of the code works fine with the actual connection string.

Well thanks for putting in your valuable time for us we will try to raise an issue with Mongoose as well and see if their team can help.

**Comment by manojsethi** [comment]
"@typegoose/typegoose": "8.3.0",
"express": "^4.17.1",
"mongoose": "5.13.13"

With the latest version of mongoose 5 and typegoose 8, the above functionality works as expected.

**Comment by hasezoey** [comment]
just tested the example code provided (minified), and could not reproduce this error

reproduction repository: https://github.com/typegoose/typegoose-testing/tree/verify633

to continue this issue, please modify the listed reproduction repository branch to reproduce the error

**Comment by manojsethi** [comment]
Thanks for the testing done. I will review your repo and get back to you asap.

**Comment by manojsethi** [comment]
Please have a look at repo

https://github.com/ankit-bc-testers/verify633.git

**Comment by manojsethi** [comment]
Hi I have made my changes but was unable to push to that particular branch. So I have created my own repo with your code

Please have a look at https://github.com/manojsethi/verify633.git

The thing is that we have prefilled data in our db and want to populate that. The connection string is there for your testing purpose which has sample data.

**Comment by hasezoey** [comment]
> I have made my changes but was unable to push to that particular branch

this is expected, if you want to make changes, either it is resolved (mostly), or create a fork and push your changes there or create a patch file (if that suits your needs better) and link it in the related issue

could you maybe re-push that repository with the actual history (or at least truncated to HEAD~3 (3 commits before your commit))? (it would make it easier to compare changes)

**Comment by manojsethi** [comment]
I forked the repo and pushed the changes over here
https://github.com/manojsethi/typegoose-testing/tree/verify633

**Comment by hasezoey** [comment]
> Shall I create a pull request?

no need

---

i just tried [your version](https://github.com/manojsethi/typegoose-testing/tree/verify633) (with the modifications of the connection string for my local install and because i dont have the sample data, i had to comment out the `findById`), and it ran without any problems (no error)

Note: from earlier comments i know / assume that the error should appear on the `databaseConfig.getModelForDb` line, not in a query (not in `findById`)

console output (with mentioned modifications):

```txt
[ 'BaseClients' ]
```

**Comment by manojsethi** [comment]
Well you can try with the connection string as well that I specifically built up for this testing only. 

See the below image for detail when error comes.

Steps to replicate.
1. You have a table A with prefilled sample data. (Not inserted with the code)
2. You have another table B in which we include a ref to A data. Considering this data is already in db and put via code for first time.
3. Now disconnect the server and I know that id already so I tried findById and re run the server without anykind of data seeding.
4. try simple findById it works
5. try findById and populate the A data it breaks.

![image](https://user-images.githubusercontent.com/5434446/143151980-8bf7dfe0-1212-47bf-8707-9ce4783ab519.png)

**Comment by hasezoey** [comment]
> The issue is in populate command when it tries to populate the nested data.

thanks, this clarifies things, and the error makes sense now

---

to explain the error:
you basically create a new model on a new model collection (in this case the connection), and dont add the lower referenced models (because in mongoose internally, they are like in the option (`ref: "hardCodedModelName"`) only referenced by a string - the model name), which means you try to populate a path with a model that does not exist on that connection - so it gives you the error that "No Schema" has been registered for model name "roles" (assume that is the name the model has in the original issue screenshot) on the current model collection (in this case the connection)

also, from what i can tell (without testing), it is a complete new `mongoose.Connection` object returned by `mongoose.connection.useDb()`
-> i just now looked at the [documentation for `useDb`](https://mongoosejs.com/docs/api.html#connection_Connection-useDb), and look at the documented return type: `«Connection» New Connection Object`
(this basically means that no models will be copied from the original connection, resulting in the models not existing when needed)

TL;DR: you try to use a model that does not exist

PS: i dont know why it worked in older mongoose version, i may look into this and open a mongoose issue as a regression / undocumented change in migration guide

**Comment by manojsethi** [comment]
I have already opened up the issue on mongoose as well and it will be good if you can escalate it on your behalf if possible. Here is the link to the issue

https://github.com/Automattic/mongoose/issues/11003

**Comment by manojsethi** [comment]
One more thing I found is that if I try to use directly mongoose 5.13.13 without typegoose 8.3.0 then also it throws same error.

```
const db = mongoose.connection.useDb("base_db");

   const client_id = db.model("clients", clients) as mongoose.Model<typeof ClientsModel>;
  
   //This Works
  const found_users = await client_id.findById(new mongoose.Types.ObjectId('619cee5091f3aaedfd5f8d80'));
  //This doesn't work
  const found_users_with_role = await client_id.findById(new mongoose.Types.ObjectId('619cee5091f3aaedfd5f8d80')).populate(['role']);

  console.log(found_users);
  console.log(found_users_with_role);
```

Whereas when I try to use it with the combination of typegoose 8.3.0, so I am just having an intuition that typegoose is doing some magic over here.

For making sure I ran the code with latest version of typegoose and mongoose like below for populate

  ```
const rolesModel = databaseConfig.getModelForDb<typeof Roles>('base_db', RolesModel);
  const found_users_with_role = await client_id
    .findById(new mongoose.Types.ObjectId('619cee5091f3aaedfd5f8d80'))
    .populate('role', {}, rolesModel);

```
//OLD populate was as below which work till typegoose 8.3.0

> .populate('role')

//New populate is as below which works with typegoose >9

> .populate('role', {}, rolesModel);

Now it works

So until 8.3.0 typegoose was doing this magic over there but that is something removed from typegoose 9

I hope this will make you reach closer to the issue.

**Comment by hasezoey** [comment]
> So until 8.3.0 typegoose was doing this magic over there but that is something removed from typegoose 9

from what i can tell and recall, typegoose 8 to 9 did not change anything related to that, what could *remotely* be related is that the order of classes compiled changes (from top-down to bottom-up), see #629

**Comment by hasezoey** [comment]
after some testing, i think the problem might be related to https://mongoosejs.com/docs/migrating_to_6.html#strictpopulate

see the following issue comment for more: https://github.com/Automattic/mongoose/issues/11003#issuecomment-977824174

**Comment by manojsethi** [comment]
Since I have tested the changes with mongoose 5 and 6 directly the way mongoose works is same I guess as I see the same errors for both versions of mongoose. With typegoose 8.3.0 mongoose 5 version start working but when we upgrade to mongoose 6 and typegoose 9 it fails.

Combination I tried
1. Mongoose 5.13.13 = Fails
2. Mongoose 5.13.13 + Typegoose 8.3.0 = Pass
3. Mongoose 6.0.12 = Fails
4. Mongoose 6.0.12 + Typegoose 9.2.0 = Fail

I need to understand more about the bottom-up approach as in our case we are not extending the classes just putting the reference.

**Comment by hasezoey** [comment]
> I need to understand more about the bottom-up approach as in our case we are not extending the classes just putting the reference.

if you dont extend, then that change is not affecting you

---

as i wrote in a earlier comment, see comment https://github.com/Automattic/mongoose/issues/11003#issuecomment-977824174 for comparison on both versions, and what might be the related change

**Comment by manojsethi** [comment]
Okay, I will wait for a reply from mongoose team and thanks for digging it further in your valuable time.

**Comment by hasezoey** [comment]
Update: the mongoose issue (https://github.com/Automattic/mongoose/issues/11003) got fixed, which will be included in mongoose 6.1.2

but typegoose wont be able to upgrade because of #647

**Comment by manojsethi** [comment]
Thanks for the update.  We will also like to wait for these errors to get fixed. I also see so many bugs included in mongoose milestone 6.1.4 and I hope mongoose team will address it in these versions.

**Comment by hasezoey** [comment]
i will now close this issue, because from what i can tell, it is fixed

mongoose shipped 6.1.2 (which included the change from what i read)
typegoose shipped 9.5.0 (which uses mongoose 6.1.3)

closing as "fixed"
